# 문제풀이

## [문제116] 근무일수가 가장 많은 10위까지 직원들의 employee_id, last_name, department_name, 근무일수를 출력해주세요.

In [1]:
# 라이브러리 불러오기 
import pandas as pd 
import datetime 
from pandas import Series, DataFrame

In [2]:
# 데이터 불러오기 
dept = pd.read_csv("c:/data/departments.csv")

# date 날짜 format 맞춰서 불러오기 
emp = pd.read_csv("c:/data/employees.csv", parse_dates = ['HIRE_DATE'], # 옵션 설정 
                 date_parser = lambda arg: pd.to_datetime(arg, format = '%Y-%m-%d'))

In [3]:
# 이름과 근무일수 뽑아주기 
x = DataFrame({'사번' : emp['EMPLOYEE_ID'],
                '이름':emp['LAST_NAME'],
                  '부서코드' : emp['DEPARTMENT_ID'],                 
               '근무일수':(pd.Timestamp.now() - emp['HIRE_DATE']).dt.days})

In [4]:
# 순위 컬럼 추가하기

x['순위'] = x['근무일수'].rank(ascending = False, method = 'dense').astype('int')

In [5]:
# rank 10위까지 출력하기 
top10 = x.loc[x['순위'] <= 10,].sort_values(by='순위', ascending = True)
top10

사번         이름   부서코드  근무일수  순위
2    102    De Haan   90.0  8075   1
103  203     Mavris   40.0  7565   2
104  204       Baer   70.0  7565   2
105  205    Higgins  110.0  7565   2
106  206      Gietz  110.0  7565   2
9    109     Faviet  100.0  7495   3
8    108  Greenberg  100.0  7494   4
14   114   Raphaely   30.0  7382   5
22   122   Kaufling   50.0  7237   6
15   115       Khoo   30.0  7220   7
0    100       King   90.0  7190   8
37   137     Ladwig   50.0  7163   9
100  200     Whalen   10.0  7098  10

In [6]:
# 부서 이름 함께 출력하기
result = pd.merge(top10, dept, left_on = '부서코드', right_on = 'DEPARTMENT_ID')[['사번','이름','부서코드','DEPARTMENT_NAME','근무일수','순위']]
result = result.rename(columns = {'DEPARTMENT_NAME' : '부서이름'})
result.sort_values(by='순위')

사번         이름   부서코드              부서이름  근무일수  순위
0   102    De Haan   90.0         Executive  8075   1
2   203     Mavris   40.0   Human Resources  7565   2
3   204       Baer   70.0  Public Relations  7565   2
4   205    Higgins  110.0        Accounting  7565   2
5   206      Gietz  110.0        Accounting  7565   2
6   109     Faviet  100.0           Finance  7495   3
7   108  Greenberg  100.0           Finance  7494   4
8   114   Raphaely   30.0        Purchasing  7382   5
10  122   Kaufling   50.0          Shipping  7237   6
9   115       Khoo   30.0        Purchasing  7220   7
1   100       King   90.0         Executive  7190   8
11  137     Ladwig   50.0          Shipping  7163   9
12  200     Whalen   10.0    Administration  7098  10

## [117] 관리자 사원의 정보를 출력해주세요.

In [31]:
c.execute('''select *
            from emp
            where employee_id in (select distinct manager_id 
                                 from emp)
            or manager_id is null''')
c.fetchall()

[(100,
  'Steven',
  'King',
  'SKING',
  '515.123.4567',
  '2003-06-17',
  'AD_PRES',
  24000,
  None,
  None,
  90.0),
 (101,
  'Neena',
  'Kochhar',
  'NKOCHHAR',
  '515.123.4568',
  '2005-09-21',
  'AD_VP',
  17000,
  None,
  100.0,
  90.0),
 (102,
  'Lex',
  'De Haan',
  'LDEHAAN',
  '515.123.4569',
  '2001-01-13',
  'AD_VP',
  17000,
  None,
  100.0,
  90.0),
 (103,
  'Alexander',
  'Hunold',
  'AHUNOLD',
  '590.423.4567',
  '2006-01-03',
  'IT_PROG',
  9000,
  None,
  102.0,
  60.0),
 (108,
  'Nancy',
  'Greenberg',
  'NGREENBE',
  '515.124.4569',
  '2002-08-17',
  'FI_MGR',
  12008,
  None,
  101.0,
  100.0),
 (114,
  'Den',
  'Raphaely',
  'DRAPHEAL',
  '515.127.4561',
  '2002-12-07',
  'PU_MAN',
  11000,
  None,
  100.0,
  30.0),
 (120,
  'Matthew',
  'Weiss',
  'MWEISS',
  '650.123.1234',
  '2004-07-18',
  'ST_MAN',
  8000,
  None,
  100.0,
  50.0),
 (121,
  'Adam',
  'Fripp',
  'AFRIPP',
  '650.123.2234',
  '2005-04-10',
  'ST_MAN',
  8200,
  None,
  100.0,
  50.0),
 (122,


In [38]:
c.execute('''select *
            from emp e
            where exists (select 'x' 
                        from emp
                        where manager_id = e.employee_id)''')
c.fetchall()

[(100,
  'Steven',
  'King',
  'SKING',
  '515.123.4567',
  '2003-06-17',
  'AD_PRES',
  24000,
  None,
  None,
  90.0),
 (101,
  'Neena',
  'Kochhar',
  'NKOCHHAR',
  '515.123.4568',
  '2005-09-21',
  'AD_VP',
  17000,
  None,
  100.0,
  90.0),
 (102,
  'Lex',
  'De Haan',
  'LDEHAAN',
  '515.123.4569',
  '2001-01-13',
  'AD_VP',
  17000,
  None,
  100.0,
  90.0),
 (103,
  'Alexander',
  'Hunold',
  'AHUNOLD',
  '590.423.4567',
  '2006-01-03',
  'IT_PROG',
  9000,
  None,
  102.0,
  60.0),
 (108,
  'Nancy',
  'Greenberg',
  'NGREENBE',
  '515.124.4569',
  '2002-08-17',
  'FI_MGR',
  12008,
  None,
  101.0,
  100.0),
 (114,
  'Den',
  'Raphaely',
  'DRAPHEAL',
  '515.127.4561',
  '2002-12-07',
  'PU_MAN',
  11000,
  None,
  100.0,
  30.0),
 (120,
  'Matthew',
  'Weiss',
  'MWEISS',
  '650.123.1234',
  '2004-07-18',
  'ST_MAN',
  8000,
  None,
  100.0,
  50.0),
 (121,
  'Adam',
  'Fripp',
  'AFRIPP',
  '650.123.2234',
  '2005-04-10',
  'ST_MAN',
  8200,
  None,
  100.0,
  50.0),
 (122,


## [118번] 관리자가 아닌 사원의 정보를 출력해주세요. 

In [36]:
# 서브쿼리 
c.execute('''select *
            from emp
            where employee_id not in (select distinct manager_id 
                                     from emp
                                     where manager_id is not null) # null을 제어하지 않으면 결과가 출력되지 않음 
            ''')
c.fetchall()

[(104,
  'Bruce',
  'Ernst',
  'BERNST',
  '590.423.4568',
  '2007-05-21',
  'IT_PROG',
  6000,
  None,
  103.0,
  60.0),
 (105,
  'David',
  'Austin',
  'DAUSTIN',
  '590.423.4569',
  '2005-06-25',
  'IT_PROG',
  4800,
  None,
  103.0,
  60.0),
 (106,
  'Valli',
  'Pataballa',
  'VPATABAL',
  '590.423.4560',
  '2006-02-05',
  'IT_PROG',
  4800,
  None,
  103.0,
  60.0),
 (107,
  'Diana',
  'Lorentz',
  'DLORENTZ',
  '590.423.5567',
  '2007-02-07',
  'IT_PROG',
  4200,
  None,
  103.0,
  60.0),
 (109,
  'Daniel',
  'Faviet',
  'DFAVIET',
  '515.124.4169',
  '2002-08-16',
  'FI_ACCOUNT',
  9000,
  None,
  108.0,
  100.0),
 (110,
  'John',
  'Chen',
  'JCHEN',
  '515.124.4269',
  '2005-09-28',
  'FI_ACCOUNT',
  8200,
  None,
  108.0,
  100.0),
 (111,
  'Ismael',
  'Sciarra',
  'ISCIARRA',
  '515.124.4369',
  '2005-09-30',
  'FI_ACCOUNT',
  7700,
  None,
  108.0,
  100.0),
 (112,
  'Jose Manuel',
  'Urman',
  'JMURMAN',
  '515.124.4469',
  '2006-03-07',
  'FI_ACCOUNT',
  7800,
  None,
  1

In [39]:
# not exists : 존재 여부만을 체크하는 연산자 
c.execute('''select *
            from emp e
            where not exists (select 'x' 
                        from emp
                        where manager_id = e.employee_id)''')
c.fetchall()

[(104,
  'Bruce',
  'Ernst',
  'BERNST',
  '590.423.4568',
  '2007-05-21',
  'IT_PROG',
  6000,
  None,
  103.0,
  60.0),
 (105,
  'David',
  'Austin',
  'DAUSTIN',
  '590.423.4569',
  '2005-06-25',
  'IT_PROG',
  4800,
  None,
  103.0,
  60.0),
 (106,
  'Valli',
  'Pataballa',
  'VPATABAL',
  '590.423.4560',
  '2006-02-05',
  'IT_PROG',
  4800,
  None,
  103.0,
  60.0),
 (107,
  'Diana',
  'Lorentz',
  'DLORENTZ',
  '590.423.5567',
  '2007-02-07',
  'IT_PROG',
  4200,
  None,
  103.0,
  60.0),
 (109,
  'Daniel',
  'Faviet',
  'DFAVIET',
  '515.124.4169',
  '2002-08-16',
  'FI_ACCOUNT',
  9000,
  None,
  108.0,
  100.0),
 (110,
  'John',
  'Chen',
  'JCHEN',
  '515.124.4269',
  '2005-09-28',
  'FI_ACCOUNT',
  8200,
  None,
  108.0,
  100.0),
 (111,
  'Ismael',
  'Sciarra',
  'ISCIARRA',
  '515.124.4369',
  '2005-09-30',
  'FI_ACCOUNT',
  7700,
  None,
  108.0,
  100.0),
 (112,
  'Jose Manuel',
  'Urman',
  'JMURMAN',
  '515.124.4469',
  '2006-03-07',
  'FI_ACCOUNT',
  7800,
  None,
  1

## [문제119] 사원들의 employee_id, salary, grade_level을 출력해주세요 

In [87]:
# 테이블 구조 파악하기 
c.execute('PRAGMA table_info(job_grades)')
c.fetchall()

[(0, 'GRADE_LEVEL', 'TEXT', 0, None, 0),
 (1, 'LOWEST_SAL', 'INTEGER', 0, None, 0),
 (2, 'HIGHEST_SAL', 'INTEGER', 0, None, 0)]

In [97]:
c.execute('''SELECT e.employee_id,
                    e.salary,
                    g.grade_level
             FROM emp e
                 JOIN job_grades g ON e.salary BETWEEN g.lowest_sal AND highest_sal''')
x = len(c.fetchall())
x

107

## [문제120] 부서이름별 총액급여를 출력해주세요

In [99]:
c.execute('''SELECT department_name,
                    sum(salary)
            FROM dept 
                RIGHT JOIN emp ON dept.department_id = emp.department_id
            GROUP BY department_name
            ORDER BY 2 DESC''')
c.fetchall()

[('Sales', 304500),
 ('Shipping', 156400),
 ('Executive', 58000),
 ('Finance', 51608),
 ('IT', 28800),
 ('Purchasing', 24900),
 ('Accounting', 20308),
 ('Marketing', 19000),
 ('Public Relations', 10000),
 (None, 7000),
 ('Human Resources', 6500),
 ('Administration', 4400)]

## [문제 121] 자신의 부서 평균 급여보다 더 많이 받는 사원들의 정보를 출력해주세요 

In [100]:
c.execute('''select e2.*
            from (select department_id, avg(salary) avgsal
                  from emp
                  group by department_id) e1, emp e2
            where e1.department_id = e2.department_id
            and e2.salary > e1.avgsal''')
c.fetchall()

[(100,
  'Steven',
  'King',
  'SKING',
  '515.123.4567',
  '2003-06-17',
  'AD_PRES',
  24000,
  None,
  None,
  90.0),
 (103,
  'Alexander',
  'Hunold',
  'AHUNOLD',
  '590.423.4567',
  '2006-01-03',
  'IT_PROG',
  9000,
  None,
  102.0,
  60.0),
 (104,
  'Bruce',
  'Ernst',
  'BERNST',
  '590.423.4568',
  '2007-05-21',
  'IT_PROG',
  6000,
  None,
  103.0,
  60.0),
 (108,
  'Nancy',
  'Greenberg',
  'NGREENBE',
  '515.124.4569',
  '2002-08-17',
  'FI_MGR',
  12008,
  None,
  101.0,
  100.0),
 (109,
  'Daniel',
  'Faviet',
  'DFAVIET',
  '515.124.4169',
  '2002-08-16',
  'FI_ACCOUNT',
  9000,
  None,
  108.0,
  100.0),
 (114,
  'Den',
  'Raphaely',
  'DRAPHEAL',
  '515.127.4561',
  '2002-12-07',
  'PU_MAN',
  11000,
  None,
  100.0,
  30.0),
 (120,
  'Matthew',
  'Weiss',
  'MWEISS',
  '650.123.1234',
  '2004-07-18',
  'ST_MAN',
  8000,
  None,
  100.0,
  50.0),
 (121,
  'Adam',
  'Fripp',
  'AFRIPP',
  '650.123.2234',
  '2005-04-10',
  'ST_MAN',
  8200,
  None,
  100.0,
  50.0),
 (1

## [문제 122] 근무 일수가 가장 많은 10위까지 직원들의 employee_id, last_name, department_name, 근무 일수를 출력하시오

In [109]:
c.execute(''' 
            select employee_id, last_name, department_name, workingday
            from (
                    select employee_id, last_name, department_id, rnk, workingday
                    from (
                        select employee_id, last_name, department_id,
                                dense_rank() over(order by workingday desc) rnk,
                                workingday
                        from (
                                select employee_id, last_name, department_id,
                                        (strftime('%s','now') - strftime('%s',date(hire_date)))/86400 workingday
                                from emp)
                    )
                    where rnk <= 10 ) e, dept d 
            where e.department_id = d.department_id 
            order by 4 desc
            ''')
c.fetchall()

[(102, 'De Haan', 'Executive', 8076),
 (203, 'Mavris', 'Human Resources', 7566),
 (204, 'Baer', 'Public Relations', 7566),
 (205, 'Higgins', 'Accounting', 7566),
 (206, 'Gietz', 'Accounting', 7566),
 (109, 'Faviet', 'Finance', 7496),
 (108, 'Greenberg', 'Finance', 7495),
 (114, 'Raphaely', 'Purchasing', 7383),
 (122, 'Kaufling', 'Shipping', 7238),
 (115, 'Khoo', 'Purchasing', 7221),
 (100, 'King', 'Executive', 7191),
 (137, 'Ladwig', 'Shipping', 7164),
 (200, 'Whalen', 'Administration', 7099)]

# 수업 내용

## sqlite 
**장점 : db가 필요가 없고, 응용프로그램 안에서 사용 가능한 가벼운 db**
- 별도의 DB서버가 필요없이 DB파일을 기초로 하여 데이터베이스를 처리하는 엔진
- 응용프로그램에서 가볍게 사용되는 데이터베이스이다.
- 현장에서 써봤냐 자주 묻는다

In [2]:
import sqlite3

In [3]:
# db를 구성하기 위한 접속! 
conn = sqlite3.connect(':memory:') #  dbms를 메모리에 만들고 버리겠다 = python 프로그램 쓰는 동안만 만들고 버리겠다
conn 

In [4]:
# cursor : SQL문 실행 메모리 영역 
c = conn.cursor()
c

In [10]:
# 오류가 나지 않으면 해당 sql문이 돌아간 것 
c.execute('create table emp(id integer, name char, sal integer)')

In [11]:
# 새로운 행 넣기 
c.execute('insert into emp(id, name, sal) values(1,"홍길동",1000)')

In [12]:
c.execute('select * from emp')

In [13]:
# 화면상에 출력하기
# 한줄씩 출력하는 것이기 때문에 두번 실행하면 아무것도 나오지 않는다 
c.fetchone()

(1, '홍길동', 1000)

In [14]:
c.execute('insert into emp(id, name, sal) values(2,"박찬호",2000)')
c.execute('select * from emp')
c.fetchone()

(1, '홍길동', 1000)

In [15]:
c.fetchone()

(2, '박찬호', 2000)

In [16]:
# 한꺼번에 출력하고 싶을 때 
c.execute('select * from emp')
c.fetchall()

[(1, '홍길동', 1000), (2, '박찬호', 2000)]

In [18]:
# transaction : 논리적으로 DML(insert, update, delete, merge)문을 한꺼번에 처리하는 작업단위 
# transaction 종료 : commit, rollback 

conn.rollback() # 커서가 아닌, coonect를 기반으로 한 rollback을 해야한다. 

In [19]:
# rollback을 해서 insert문이 취소가 됨 

c.execute('select * from emp')
c.fetchall()

[]

In [20]:
# 다시한번 정보 insert 

c.execute('insert into emp(id, name, sal) values(1,"홍길동",1000)')
c.execute('insert into emp(id, name, sal) values(2,"박찬호",2000)')

In [22]:
# 다시 한번 조회하기

c.execute('select * from emp')
c.fetchall()

[(1, '홍길동', 1000), (2, '박찬호', 2000)]

In [23]:
conn.commit() # 영구히(memory 안에서만) 저장됨. . 

In [25]:
# 이미 트랜젝션을 종료되어서 rollback을 해도 소용이 없다 
conn.rollback()

c.execute('select * from emp')
c.fetchall()

[(1, '홍길동', 1000), (2, '박찬호', 2000)]

In [26]:
# sqllite을 종료할 때는 cursor -> conn 종료해야 함. 
## 위에서 insert한 데이터들은 memory가 종료되어 사라지게 됨 
c.close()
conn.close()

In [27]:
# 다시 한번 조회하면 ProgrammingError가 뜨게 됨 

c.execute('select * from emp')
c.fetchall()

ProgrammingError: Cannot operate on a closed cursor.

In [29]:
conn = sqlite3.connect(':memory:')
c = conn.cursor()
c.execute('select * from emp') # 테이블이 없어서 아래와 같은 에러가 뜸 

OperationalError: no such table: emp

In [30]:
c.close()
conn.close()

## DBMS처럼 영구히 저장하고 싶을 때?=> 메모리가 아닌 특정한 파일로 저장하기 
- sqlite에서 수행한 데이터를 영구히 보존하는 방법
- 현장에서 사용하는 방법

In [50]:
conn = sqlite3.connect('c:/data/insa.db')

In [51]:
conn = sqlite3.connect('c:/data/insa.db')
c = conn.cursor()
c.execute('create table emp(id integer, name char, sal integer)')
c.execute('insert into emp(id,name,sal) values(1,"홍길동",1000)')
c.execute('insert into emp(id,name,sal) values(2,"박찬호",2000)')
c.execute('select * from emp')
c.fetchall()

OperationalError: table emp already exists

In [52]:
c.close()
conn.close()

In [53]:
conn = sqlite3.connect('c:/data/insa.db') # 데이터베이스 생성, 접속, 열기  
c = conn.cursor() # SQL문을 실행하기 위해서 메모리 할당 
c.execute('select * from emp')
c.fetchall()

[(100,
  'Steven',
  'King',
  'SKING',
  '515.123.4567',
  '2003-06-17',
  'AD_PRES',
  24000,
  None,
  None,
  90.0),
 (101,
  'Neena',
  'Kochhar',
  'NKOCHHAR',
  '515.123.4568',
  '2005-09-21',
  'AD_VP',
  17000,
  None,
  100.0,
  90.0),
 (102,
  'Lex',
  'De Haan',
  'LDEHAAN',
  '515.123.4569',
  '2001-01-13',
  'AD_VP',
  17000,
  None,
  100.0,
  90.0),
 (103,
  'Alexander',
  'Hunold',
  'AHUNOLD',
  '590.423.4567',
  '2006-01-03',
  'IT_PROG',
  9000,
  None,
  102.0,
  60.0),
 (104,
  'Bruce',
  'Ernst',
  'BERNST',
  '590.423.4568',
  '2007-05-21',
  'IT_PROG',
  6000,
  None,
  103.0,
  60.0),
 (105,
  'David',
  'Austin',
  'DAUSTIN',
  '590.423.4569',
  '2005-06-25',
  'IT_PROG',
  4800,
  None,
  103.0,
  60.0),
 (106,
  'Valli',
  'Pataballa',
  'VPATABAL',
  '590.423.4560',
  '2006-02-05',
  'IT_PROG',
  4800,
  None,
  103.0,
  60.0),
 (107,
  'Diana',
  'Lorentz',
  'DLORENTZ',
  '590.423.5567',
  '2007-02-07',
  'IT_PROG',
  4200,
  None,
  103.0,
  60.0),
 (10

In [56]:
# insa라는 db안에 어떤 테이블이 있는가?를 알아보기

c.execute('select name from sqlite_master where type = "table"')
c.fetchall()

[('emp',), ('dept',), ('job_grades',)]

In [58]:
# t테이블의 구조를 확인하기 
c.execute('PRAGMA table_info(emp)')
c.fetchall()

[(0, 'EMPLOYEE_ID', 'INTEGER', 0, None, 0),
 (1, 'FIRST_NAME', 'TEXT', 0, None, 0),
 (2, 'LAST_NAME', 'TEXT', 0, None, 0),
 (3, 'EMAIL', 'TEXT', 0, None, 0),
 (4, 'PHONE_NUMBER', 'TEXT', 0, None, 0),
 (5, 'HIRE_DATE', 'TEXT', 0, None, 0),
 (6, 'JOB_ID', 'TEXT', 0, None, 0),
 (7, 'SALARY', 'INTEGER', 0, None, 0),
 (8, 'COMMISSION_PCT', 'REAL', 0, None, 0),
 (9, 'MANAGER_ID', 'REAL', 0, None, 0),
 (10, 'DEPARTMENT_ID', 'REAL', 0, None, 0)]

In [59]:
# DDL : 테이블이 영구히 삭제됨 
c.execute('drop table emp')

In [60]:
# 조회했을 때 없음 
c.execute('select name from sqlite_master where type = "table"')
c.fetchall()

[('dept',), ('job_grades',)]

## pandas dataframe을 sqlite table로 이관 작업하기
- pandas 제공 

In [61]:
# 데이터 열기 
data = pd.read_csv("c:/data/employees.csv")

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 107 entries, 0 to 106
Data columns (total 11 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   EMPLOYEE_ID     107 non-null    int64  
 1   FIRST_NAME      107 non-null    object 
 2   LAST_NAME       107 non-null    object 
 3   EMAIL           107 non-null    object 
 4   PHONE_NUMBER    107 non-null    object 
 5   HIRE_DATE       107 non-null    object 
 6   JOB_ID          107 non-null    object 
 7   SALARY          107 non-null    int64  
 8   COMMISSION_PCT  35 non-null     float64
 9   MANAGER_ID      106 non-null    float64
 10  DEPARTMENT_ID   106 non-null    float64
dtypes: float64(3), int64(2), object(6)
memory usage: 9.3+ KB


In [62]:
# 데이터 load하기 
data.to_sql("emp",conn, index = False) 

107

In [63]:
c.execute('select name from sqlite_master where type = "table"')
c.fetchall()

[('dept',), ('job_grades',), ('emp',)]

In [64]:
c.execute('PRAGMA table_info(emp)')
c.fetchall()

[(0, 'EMPLOYEE_ID', 'INTEGER', 0, None, 0),
 (1, 'FIRST_NAME', 'TEXT', 0, None, 0),
 (2, 'LAST_NAME', 'TEXT', 0, None, 0),
 (3, 'EMAIL', 'TEXT', 0, None, 0),
 (4, 'PHONE_NUMBER', 'TEXT', 0, None, 0),
 (5, 'HIRE_DATE', 'TEXT', 0, None, 0),
 (6, 'JOB_ID', 'TEXT', 0, None, 0),
 (7, 'SALARY', 'INTEGER', 0, None, 0),
 (8, 'COMMISSION_PCT', 'REAL', 0, None, 0),
 (9, 'MANAGER_ID', 'REAL', 0, None, 0),
 (10, 'DEPARTMENT_ID', 'REAL', 0, None, 0)]

In [65]:
c.execute('select * from emp') # 데이터를 확인 가능하다 
c.fetchall()

[(100,
  'Steven',
  'King',
  'SKING',
  '515.123.4567',
  '2003-06-17',
  'AD_PRES',
  24000,
  None,
  None,
  90.0),
 (101,
  'Neena',
  'Kochhar',
  'NKOCHHAR',
  '515.123.4568',
  '2005-09-21',
  'AD_VP',
  17000,
  None,
  100.0,
  90.0),
 (102,
  'Lex',
  'De Haan',
  'LDEHAAN',
  '515.123.4569',
  '2001-01-13',
  'AD_VP',
  17000,
  None,
  100.0,
  90.0),
 (103,
  'Alexander',
  'Hunold',
  'AHUNOLD',
  '590.423.4567',
  '2006-01-03',
  'IT_PROG',
  9000,
  None,
  102.0,
  60.0),
 (104,
  'Bruce',
  'Ernst',
  'BERNST',
  '590.423.4568',
  '2007-05-21',
  'IT_PROG',
  6000,
  None,
  103.0,
  60.0),
 (105,
  'David',
  'Austin',
  'DAUSTIN',
  '590.423.4569',
  '2005-06-25',
  'IT_PROG',
  4800,
  None,
  103.0,
  60.0),
 (106,
  'Valli',
  'Pataballa',
  'VPATABAL',
  '590.423.4560',
  '2006-02-05',
  'IT_PROG',
  4800,
  None,
  103.0,
  60.0),
 (107,
  'Diana',
  'Lorentz',
  'DLORENTZ',
  '590.423.5567',
  '2007-02-07',
  'IT_PROG',
  4200,
  None,
  103.0,
  60.0),
 (10

In [5]:
c.close()
conn.close()

In [6]:
# 다시 열어서 확인해도 잘 살아있다

conn = sqlite3.connect('c:/data/insa.db')
c = conn.cursor()
c.execute('select * from emp')
c.fetchall()

[(100,
  'Steven',
  'King',
  'SKING',
  '515.123.4567',
  '2003-06-17',
  'AD_PRES',
  24000,
  None,
  None,
  90.0),
 (101,
  'Neena',
  'Kochhar',
  'NKOCHHAR',
  '515.123.4568',
  '2005-09-21',
  'AD_VP',
  17000,
  None,
  100.0,
  90.0),
 (102,
  'Lex',
  'De Haan',
  'LDEHAAN',
  '515.123.4569',
  '2001-01-13',
  'AD_VP',
  17000,
  None,
  100.0,
  90.0),
 (103,
  'Alexander',
  'Hunold',
  'AHUNOLD',
  '590.423.4567',
  '2006-01-03',
  'IT_PROG',
  9000,
  None,
  102.0,
  60.0),
 (104,
  'Bruce',
  'Ernst',
  'BERNST',
  '590.423.4568',
  '2007-05-21',
  'IT_PROG',
  6000,
  None,
  103.0,
  60.0),
 (105,
  'David',
  'Austin',
  'DAUSTIN',
  '590.423.4569',
  '2005-06-25',
  'IT_PROG',
  4800,
  None,
  103.0,
  60.0),
 (106,
  'Valli',
  'Pataballa',
  'VPATABAL',
  '590.423.4560',
  '2006-02-05',
  'IT_PROG',
  4800,
  None,
  103.0,
  60.0),
 (107,
  'Diana',
  'Lorentz',
  'DLORENTZ',
  '590.423.5567',
  '2007-02-07',
  'IT_PROG',
  4200,
  None,
  103.0,
  60.0),
 (10

## sqlite table을 pandas DataFrame로 이관작업

In [69]:
emp_new = pd.read_sql_query("select employee_id, last_name, salary, department_id from emp", conn)
emp_new # DataFrame으로 잘 불러와졌다. !! 

EMPLOYEE_ID LAST_NAME  SALARY  DEPARTMENT_ID
0            100      King   24000           90.0
1            101   Kochhar   17000           90.0
2            102   De Haan   17000           90.0
3            103    Hunold    9000           60.0
4            104     Ernst    6000           60.0
..           ...       ...     ...            ...
102          202       Fay    6000           20.0
103          203    Mavris    6500           40.0
104          204      Baer   10000           70.0
105          205   Higgins   12008          110.0
106          206     Gietz    8300          110.0

[107 rows x 4 columns]

In [70]:
emp_new.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 107 entries, 0 to 106
Data columns (total 4 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   EMPLOYEE_ID    107 non-null    int64  
 1   LAST_NAME      107 non-null    object 
 2   SALARY         107 non-null    int64  
 3   DEPARTMENT_ID  106 non-null    float64
dtypes: float64(1), int64(2), object(1)
memory usage: 3.5+ KB


In [73]:
# where절 사용해보기 
c.execute('select * from emp where last_name = "King"')
c.fetchall()

[(100,
  'Steven',
  'King',
  'SKING',
  '515.123.4567',
  '2003-06-17',
  'AD_PRES',
  24000,
  None,
  None,
  90.0),
 (156,
  'Janette',
  'King',
  'JKING',
  '011.44.1345.429268',
  '2004-01-30',
  'SA_REP',
  10000,
  0.35,
  146.0,
  80.0)]

In [74]:
# 소문자로 형변환
# 대소문자를 구분한다
c.execute('select * from emp where lower(last_name) = "king"')
c.fetchall()

[(100,
  'Steven',
  'King',
  'SKING',
  '515.123.4567',
  '2003-06-17',
  'AD_PRES',
  24000,
  None,
  None,
  90.0),
 (156,
  'Janette',
  'King',
  'JKING',
  '011.44.1345.429268',
  '2004-01-30',
  'SA_REP',
  10000,
  0.35,
  146.0,
  80.0)]

In [77]:
# 대문자로 형을 바꾸기 
c.execute('select * from emp where upper(last_name) = "KING"')
c.fetchall()

[(100,
  'Steven',
  'King',
  'SKING',
  '515.123.4567',
  '2003-06-17',
  'AD_PRES',
  24000,
  None,
  None,
  90.0),
 (156,
  'Janette',
  'King',
  'JKING',
  '011.44.1345.429268',
  '2004-01-30',
  'SA_REP',
  10000,
  0.35,
  146.0,
  80.0)]

In [79]:
c.execute('select last_name, substr(last_name,1,2),substr(last_name,-2,2) from emp') # 앞에서 두번째, 뒤에서 두번째 글자 추출 
c.fetchall()

[('King', 'Ki', 'ng'),
 ('Kochhar', 'Ko', 'ar'),
 ('De Haan', 'De', 'an'),
 ('Hunold', 'Hu', 'ld'),
 ('Ernst', 'Er', 'st'),
 ('Austin', 'Au', 'in'),
 ('Pataballa', 'Pa', 'la'),
 ('Lorentz', 'Lo', 'tz'),
 ('Greenberg', 'Gr', 'rg'),
 ('Faviet', 'Fa', 'et'),
 ('Chen', 'Ch', 'en'),
 ('Sciarra', 'Sc', 'ra'),
 ('Urman', 'Ur', 'an'),
 ('Popp', 'Po', 'pp'),
 ('Raphaely', 'Ra', 'ly'),
 ('Khoo', 'Kh', 'oo'),
 ('Baida', 'Ba', 'da'),
 ('Tobias', 'To', 'as'),
 ('Himuro', 'Hi', 'ro'),
 ('Colmenares', 'Co', 'es'),
 ('Weiss', 'We', 'ss'),
 ('Fripp', 'Fr', 'pp'),
 ('Kaufling', 'Ka', 'ng'),
 ('Vollman', 'Vo', 'an'),
 ('Mourgos', 'Mo', 'os'),
 ('Nayer', 'Na', 'er'),
 ('Mikkilineni', 'Mi', 'ni'),
 ('Landry', 'La', 'ry'),
 ('Markle', 'Ma', 'le'),
 ('Bissot', 'Bi', 'ot'),
 ('Atkinson', 'At', 'on'),
 ('Marlow', 'Ma', 'ow'),
 ('Olson', 'Ol', 'on'),
 ('Mallin', 'Ma', 'in'),
 ('Rogers', 'Ro', 'rs'),
 ('Gee', 'Ge', 'ee'),
 ('Philtanker', 'Ph', 'er'),
 ('Ladwig', 'La', 'ig'),
 ('Stiles', 'St', 'es'),
 ('Seo', 'Se

In [82]:
# 첫글자는 대문자, 뒤부터는 소문자로 하는 INITCAP 구현하기 
c.execute('select upper(substr(last_name,1,1))|| lower(substr(last_name,2)) from emp')
c.fetchall()

[('King',),
 ('Kochhar',),
 ('De haan',),
 ('Hunold',),
 ('Ernst',),
 ('Austin',),
 ('Pataballa',),
 ('Lorentz',),
 ('Greenberg',),
 ('Faviet',),
 ('Chen',),
 ('Sciarra',),
 ('Urman',),
 ('Popp',),
 ('Raphaely',),
 ('Khoo',),
 ('Baida',),
 ('Tobias',),
 ('Himuro',),
 ('Colmenares',),
 ('Weiss',),
 ('Fripp',),
 ('Kaufling',),
 ('Vollman',),
 ('Mourgos',),
 ('Nayer',),
 ('Mikkilineni',),
 ('Landry',),
 ('Markle',),
 ('Bissot',),
 ('Atkinson',),
 ('Marlow',),
 ('Olson',),
 ('Mallin',),
 ('Rogers',),
 ('Gee',),
 ('Philtanker',),
 ('Ladwig',),
 ('Stiles',),
 ('Seo',),
 ('Patel',),
 ('Rajs',),
 ('Davies',),
 ('Matos',),
 ('Vargas',),
 ('Russell',),
 ('Partners',),
 ('Errazuriz',),
 ('Cambrault',),
 ('Zlotkey',),
 ('Tucker',),
 ('Bernstein',),
 ('Hall',),
 ('Olsen',),
 ('Cambrault',),
 ('Tuvault',),
 ('King',),
 ('Sully',),
 ('Mcewen',),
 ('Smith',),
 ('Doran',),
 ('Sewall',),
 ('Vishney',),
 ('Greene',),
 ('Marvins',),
 ('Lee',),
 ('Ande',),
 ('Banda',),
 ('Ozer',),
 ('Bloom',),
 ('Fox',),
 

In [87]:
# 다음줄로 떨어트리고 싶을 때는 따옴표 세개를 사용해보자

c.execute('''select upper(substr(last_name,1,1))|| lower(substr(last_name,2)) , length(last_name)
              from emp
              where department_id = 30''')
c.fetchall()

[('Raphaely', 8),
 ('Khoo', 4),
 ('Baida', 5),
 ('Tobias', 6),
 ('Himuro', 6),
 ('Colmenares', 10)]

In [88]:
c.execute('select 1 + 2 ') # dual을 할 필요가 없음
c.fetchall()

[(3,)]

In [92]:
c.execute("select replace('데이터분석가','분석가','과학자')") # 특정 문자 치환 
c.fetchall()

[('데이터과학자',)]

In [93]:
c.execute('select trim("  데이터분석가  ")') # 공백 제거 
c.fetchall()

[('데이터분석가',)]

In [95]:
c.execute('select instr("james@naver.com","@")') # 특정 문자 위치 찾기 
c.fetchall()

[(6,)]

In [9]:
c.execute('select 1+2, 1-2, 2*3, 10/3, 10%3') # 사칙연산은 dual없이 가능함 
c.fetchall()

[(3, -1, 6, 3, 1)]

In [14]:
c.execute('select round(45,926), round(45.926, 2), round(45.926, -1)') # -1, -2로 일의자리 십의자리 반올림은 구현이 되지 못했다. 
c.fetchall()

[(45.0, 45.93, 46.0)]

In [16]:
c.execute('select trunc(45.926)') # trunc 함수 또한 구현이 되지 않음
c.fetchall()

OperationalError: no such function: truncation

In [18]:
c.execute('select last_name, ifnull(salary * 12 + commission_pct,salary) from emp') # nvl 함수와 동일하다 
c.fetchall()

[('King', 24000),
 ('Kochhar', 17000),
 ('De Haan', 17000),
 ('Hunold', 9000),
 ('Ernst', 6000),
 ('Austin', 4800),
 ('Pataballa', 4800),
 ('Lorentz', 4200),
 ('Greenberg', 12008),
 ('Faviet', 9000),
 ('Chen', 8200),
 ('Sciarra', 7700),
 ('Urman', 7800),
 ('Popp', 6900),
 ('Raphaely', 11000),
 ('Khoo', 3100),
 ('Baida', 2900),
 ('Tobias', 2800),
 ('Himuro', 2600),
 ('Colmenares', 2500),
 ('Weiss', 8000),
 ('Fripp', 8200),
 ('Kaufling', 7900),
 ('Vollman', 6500),
 ('Mourgos', 5800),
 ('Nayer', 3200),
 ('Mikkilineni', 2700),
 ('Landry', 2400),
 ('Markle', 2200),
 ('Bissot', 3300),
 ('Atkinson', 2800),
 ('Marlow', 2500),
 ('Olson', 2100),
 ('Mallin', 3300),
 ('Rogers', 2900),
 ('Gee', 2400),
 ('Philtanker', 2200),
 ('Ladwig', 3600),
 ('Stiles', 3200),
 ('Seo', 2700),
 ('Patel', 2500),
 ('Rajs', 3500),
 ('Davies', 3100),
 ('Matos', 2600),
 ('Vargas', 2500),
 ('Russell', 168000.4),
 ('Partners', 162000.3),
 ('Errazuriz', 144000.3),
 ('Cambrault', 132000.3),
 ('Zlotkey', 126000.2),
 ('Tucker

In [22]:
c.execute('select last_name, length(last_name), nullif(length(last_name),5) from emp') # null값을 만드는 함수 
c.fetchall()

[('King', 4, 4),
 ('Kochhar', 7, 7),
 ('De Haan', 7, 7),
 ('Hunold', 6, 6),
 ('Ernst', 5, None),
 ('Austin', 6, 6),
 ('Pataballa', 9, 9),
 ('Lorentz', 7, 7),
 ('Greenberg', 9, 9),
 ('Faviet', 6, 6),
 ('Chen', 4, 4),
 ('Sciarra', 7, 7),
 ('Urman', 5, None),
 ('Popp', 4, 4),
 ('Raphaely', 8, 8),
 ('Khoo', 4, 4),
 ('Baida', 5, None),
 ('Tobias', 6, 6),
 ('Himuro', 6, 6),
 ('Colmenares', 10, 10),
 ('Weiss', 5, None),
 ('Fripp', 5, None),
 ('Kaufling', 8, 8),
 ('Vollman', 7, 7),
 ('Mourgos', 7, 7),
 ('Nayer', 5, None),
 ('Mikkilineni', 11, 11),
 ('Landry', 6, 6),
 ('Markle', 6, 6),
 ('Bissot', 6, 6),
 ('Atkinson', 8, 8),
 ('Marlow', 6, 6),
 ('Olson', 5, None),
 ('Mallin', 6, 6),
 ('Rogers', 6, 6),
 ('Gee', 3, 3),
 ('Philtanker', 10, 10),
 ('Ladwig', 6, 6),
 ('Stiles', 6, 6),
 ('Seo', 3, 3),
 ('Patel', 5, None),
 ('Rajs', 4, 4),
 ('Davies', 6, 6),
 ('Matos', 5, None),
 ('Vargas', 6, 6),
 ('Russell', 7, 7),
 ('Partners', 8, 8),
 ('Errazuriz', 9, 9),
 ('Cambrault', 9, 9),
 ('Zlotkey', 7, 7),
 

In [20]:
c.execute('select last_name, nullif(salary * 12 + commission_pct,salary) from emp') # nvl 함수와 동일하다 
c.fetchall()

[('King', None),
 ('Kochhar', None),
 ('De Haan', None),
 ('Hunold', None),
 ('Ernst', None),
 ('Austin', None),
 ('Pataballa', None),
 ('Lorentz', None),
 ('Greenberg', None),
 ('Faviet', None),
 ('Chen', None),
 ('Sciarra', None),
 ('Urman', None),
 ('Popp', None),
 ('Raphaely', None),
 ('Khoo', None),
 ('Baida', None),
 ('Tobias', None),
 ('Himuro', None),
 ('Colmenares', None),
 ('Weiss', None),
 ('Fripp', None),
 ('Kaufling', None),
 ('Vollman', None),
 ('Mourgos', None),
 ('Nayer', None),
 ('Mikkilineni', None),
 ('Landry', None),
 ('Markle', None),
 ('Bissot', None),
 ('Atkinson', None),
 ('Marlow', None),
 ('Olson', None),
 ('Mallin', None),
 ('Rogers', None),
 ('Gee', None),
 ('Philtanker', None),
 ('Ladwig', None),
 ('Stiles', None),
 ('Seo', None),
 ('Patel', None),
 ('Rajs', None),
 ('Davies', None),
 ('Matos', None),
 ('Vargas', None),
 ('Russell', 168000.4),
 ('Partners', 162000.3),
 ('Errazuriz', 144000.3),
 ('Cambrault', 132000.3),
 ('Zlotkey', 126000.2),
 ('Tucker', 12

In [24]:
c.execute('select distinct department_id from emp order by 1')
c.fetchall()

[(None,),
 (10.0,),
 (20.0,),
 (30.0,),
 (40.0,),
 (50.0,),
 (60.0,),
 (70.0,),
 (80.0,),
 (90.0,),
 (100.0,),
 (110.0,)]

In [27]:
c.execute('select * from emp where last_name like "K%"') # 대문자 K로 시작되는 이름을 가진 사원 정보 출력 
c.fetchall()

[(100,
  'Steven',
  'King',
  'SKING',
  '515.123.4567',
  '2003-06-17',
  'AD_PRES',
  24000,
  None,
  None,
  90.0),
 (101,
  'Neena',
  'Kochhar',
  'NKOCHHAR',
  '515.123.4568',
  '2005-09-21',
  'AD_VP',
  17000,
  None,
  100.0,
  90.0),
 (115,
  'Alexander',
  'Khoo',
  'AKHOO',
  '515.127.4562',
  '2003-05-18',
  'PU_CLERK',
  3100,
  None,
  114.0,
  30.0),
 (122,
  'Payam',
  'Kaufling',
  'PKAUFLIN',
  '650.123.3234',
  '2003-05-01',
  'ST_MAN',
  7900,
  None,
  100.0,
  50.0),
 (156,
  'Janette',
  'King',
  'JKING',
  '011.44.1345.429268',
  '2004-01-30',
  'SA_REP',
  10000,
  0.35,
  146.0,
  80.0),
 (173,
  'Sundita',
  'Kumar',
  'SKUMAR',
  '011.44.1343.329268',
  '2008-04-21',
  'SA_REP',
  6100,
  0.1,
  148.0,
  80.0)]

In [40]:
c.execute('select date("now"), datetime("now"), datetime("now","localtime")')
c.fetchall()

[('2023-02-23', '2023-02-23 05:07:59', '2023-02-23 14:07:59')]

In [42]:
c.execute('select date("now","100 day")') # 날짜 + 일수 
c.fetchall() 

[('2023-06-03',)]

In [44]:
c.execute('select date("now","-100 month")') # 날짜 - 달수 
c.fetchall() 

[('2014-10-23',)]

In [45]:
c.execute('select date("now","1 month","2 day")')
c.fetchall() 

[('2023-03-25',)]

In [49]:
c.execute('select datetime("now","localtime", "1 hour", "10 minute" )')
c.fetchall()

[('2023-02-23 15:52:16',)]

In [57]:
c.execute('select date ("2021-09-29","weekday 3")') # 가장 가까운 요일 0~6(일~토)
c.fetchall()

[('2021-09-29',)]

In [60]:
# case문 문법 동일함 
c.execute('''select last_name, salary, commission_pct,
                    case when commission_pct is null then salary * 12 
                         else (salary * 12) + (salary * 12 * commission_pct)
                    end as comm_pct
            from emp''')
c.fetchall()

[('King', 24000, None, 288000),
 ('Kochhar', 17000, None, 204000),
 ('De Haan', 17000, None, 204000),
 ('Hunold', 9000, None, 108000),
 ('Ernst', 6000, None, 72000),
 ('Austin', 4800, None, 57600),
 ('Pataballa', 4800, None, 57600),
 ('Lorentz', 4200, None, 50400),
 ('Greenberg', 12008, None, 144096),
 ('Faviet', 9000, None, 108000),
 ('Chen', 8200, None, 98400),
 ('Sciarra', 7700, None, 92400),
 ('Urman', 7800, None, 93600),
 ('Popp', 6900, None, 82800),
 ('Raphaely', 11000, None, 132000),
 ('Khoo', 3100, None, 37200),
 ('Baida', 2900, None, 34800),
 ('Tobias', 2800, None, 33600),
 ('Himuro', 2600, None, 31200),
 ('Colmenares', 2500, None, 30000),
 ('Weiss', 8000, None, 96000),
 ('Fripp', 8200, None, 98400),
 ('Kaufling', 7900, None, 94800),
 ('Vollman', 6500, None, 78000),
 ('Mourgos', 5800, None, 69600),
 ('Nayer', 3200, None, 38400),
 ('Mikkilineni', 2700, None, 32400),
 ('Landry', 2400, None, 28800),
 ('Markle', 2200, None, 26400),
 ('Bissot', 3300, None, 39600),
 ('Atkinson', 280

In [68]:
c.execute('''select count(*),
                    count(case substr(hire_date,1,4) when '2005' then 1 end),
                    count(case substr(hire_date,1,4) when '2006' then 1 end),
                    count(case substr(hire_date,1,4) when '2007' then 1 end)
            from emp''')
c.fetchall()

[(107, 29, 24, 19)]

In [72]:
c.execute('''select count(*),
                    count(case strftime("%Y",date(hire_date)) when '2005' then 1 end),
                    count(case substr(hire_date,1,4) when '2006' then 1 end),
                    count(case substr(hire_date,1,4) when '2007' then 1 end)
            from emp''')
c.fetchall()

[(107, 29, 24, 19)]

In [75]:
c.execute("select sum(salary), round(avg(salary),2), max(salary), min(salary) from emp")
c.fetchall()

[(691416, 6461.83, 24000, 2100)]

In [77]:
c.execute('''select department_id, 
                    sum(salary)
            from emp
            group by department_id''')
c.fetchall()

[(None, 7000),
 (10.0, 4400),
 (20.0, 19000),
 (30.0, 24900),
 (40.0, 6500),
 (50.0, 156400),
 (60.0, 28800),
 (70.0, 10000),
 (80.0, 304500),
 (90.0, 58000),
 (100.0, 51608),
 (110.0, 20308)]

## sqlite join

In [79]:
import pandas as pd 

data = pd.read_csv("c:/data/departments.csv")
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 27 entries, 0 to 26
Data columns (total 4 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   DEPARTMENT_ID    27 non-null     int64  
 1   DEPARTMENT_NAME  27 non-null     object 
 2   MANAGER_ID       11 non-null     float64
 3   LOCATION_ID      27 non-null     int64  
dtypes: float64(1), int64(2), object(1)
memory usage: 992.0+ bytes


In [81]:
# sqlite로 이동하기
data.to_sql('dept',conn,index=False)

ValueError: Table 'dept' already exists.

In [83]:
c.execute('select * from dept')
c.fetchall()

[(10, 'Administration', 200.0, 1700),
 (20, 'Marketing', 201.0, 1800),
 (30, 'Purchasing', 114.0, 1700),
 (40, 'Human Resources', 203.0, 2400),
 (50, 'Shipping', 121.0, 1500),
 (60, 'IT', 103.0, 1400),
 (70, 'Public Relations', 204.0, 2700),
 (80, 'Sales', 145.0, 2500),
 (90, 'Executive', 100.0, 1700),
 (100, 'Finance', 108.0, 1700),
 (110, 'Accounting', 205.0, 1700),
 (120, 'Treasury', None, 1700),
 (130, 'Corporate Tax', None, 1700),
 (140, 'Control And Credit', None, 1700),
 (150, 'Shareholder Services', None, 1700),
 (160, 'Benefits', None, 1700),
 (170, 'Manufacturing', None, 1700),
 (180, 'Construction', None, 1700),
 (190, 'Contracting', None, 1700),
 (200, 'Operations', None, 1700),
 (210, 'IT Support', None, 1700),
 (220, 'NOC', None, 1700),
 (230, 'IT Helpdesk', None, 1700),
 (240, 'Government Sales', None, 1700),
 (250, 'Retail Sales', None, 1700),
 (260, 'Recruiting', None, 1700),
 (270, 'Payroll', None, 1700)]

In [85]:
# cartisian product가 발생한다. 

c.execute('''select employee_id, last_name, department_name 
            from emp, dept''')
c.fetchall()

[(100, 'King', 'Administration'),
 (100, 'King', 'Marketing'),
 (100, 'King', 'Purchasing'),
 (100, 'King', 'Human Resources'),
 (100, 'King', 'Shipping'),
 (100, 'King', 'IT'),
 (100, 'King', 'Public Relations'),
 (100, 'King', 'Sales'),
 (100, 'King', 'Executive'),
 (100, 'King', 'Finance'),
 (100, 'King', 'Accounting'),
 (100, 'King', 'Treasury'),
 (100, 'King', 'Corporate Tax'),
 (100, 'King', 'Control And Credit'),
 (100, 'King', 'Shareholder Services'),
 (100, 'King', 'Benefits'),
 (100, 'King', 'Manufacturing'),
 (100, 'King', 'Construction'),
 (100, 'King', 'Contracting'),
 (100, 'King', 'Operations'),
 (100, 'King', 'IT Support'),
 (100, 'King', 'NOC'),
 (100, 'King', 'IT Helpdesk'),
 (100, 'King', 'Government Sales'),
 (100, 'King', 'Retail Sales'),
 (100, 'King', 'Recruiting'),
 (100, 'King', 'Payroll'),
 (101, 'Kochhar', 'Administration'),
 (101, 'Kochhar', 'Marketing'),
 (101, 'Kochhar', 'Purchasing'),
 (101, 'Kochhar', 'Human Resources'),
 (101, 'Kochhar', 'Shipping'),
 (

In [86]:
# equi join, simple join, inner join (카티시안 곱이 발생하지 않도록)
c.execute('''select employee_id, last_name, department_name 
            from emp e, dept d
             where e.department_id = d.department_id''')
c.fetchall()

[(100, 'King', 'Executive'),
 (101, 'Kochhar', 'Executive'),
 (102, 'De Haan', 'Executive'),
 (103, 'Hunold', 'IT'),
 (104, 'Ernst', 'IT'),
 (105, 'Austin', 'IT'),
 (106, 'Pataballa', 'IT'),
 (107, 'Lorentz', 'IT'),
 (108, 'Greenberg', 'Finance'),
 (109, 'Faviet', 'Finance'),
 (110, 'Chen', 'Finance'),
 (111, 'Sciarra', 'Finance'),
 (112, 'Urman', 'Finance'),
 (113, 'Popp', 'Finance'),
 (114, 'Raphaely', 'Purchasing'),
 (115, 'Khoo', 'Purchasing'),
 (116, 'Baida', 'Purchasing'),
 (117, 'Tobias', 'Purchasing'),
 (118, 'Himuro', 'Purchasing'),
 (119, 'Colmenares', 'Purchasing'),
 (120, 'Weiss', 'Shipping'),
 (121, 'Fripp', 'Shipping'),
 (122, 'Kaufling', 'Shipping'),
 (123, 'Vollman', 'Shipping'),
 (124, 'Mourgos', 'Shipping'),
 (125, 'Nayer', 'Shipping'),
 (126, 'Mikkilineni', 'Shipping'),
 (127, 'Landry', 'Shipping'),
 (128, 'Markle', 'Shipping'),
 (129, 'Bissot', 'Shipping'),
 (130, 'Atkinson', 'Shipping'),
 (131, 'Marlow', 'Shipping'),
 (132, 'Olson', 'Shipping'),
 (133, 'Mallin', 'S